sqlalchemy_exasol/merge.py

```diff
@@ -11,115 +16,140 @@
 
 
 class Merge(UpdateBase):
-
     __visit_name__ = "merge"
 
+    class Delete(UpdateBase):
+        def __init__(self, where=None):
+            self.where = where
+
     def __init__(self, target_table, source_expr, on):
-        self._target_table = target_table
+        self.table = target_table
         self._source_expr = source_expr
         self._on = on
-        self._on_columns = []
-        elements_to_check = list(on.get_children())
-        for e in elements_to_check:
-            if isinstance(e, Column):
-                if e.table == self._target_table:
-                    self._on_columns.append(e)
-            else:
-                elements_to_check.extend(e.get_children())
+        self._on_columns = list(
+            element
+            for element in on.get_children()
+            if isinstance(element, Column) and element.table == self.table
+        )
         self._merge_update_values = None
         self._update_where = None
         self._merge_insert_values = None
         self._insert_where = None
         self._merge_delete = False
         self._delete_where = None
+        self._delete = None
 
-    def _get_source_cols(self):
-        source_cols = {}
-        elements_to_check = list(self._source_expr.get_children())
-        for e in elements_to_check:
-            if isinstance(e, ColumnClause):
-                source_cols[e.name] = e
-        return source_cols
+    def _source_columns(self):
+        elements = list(
+            element
+            for element in self._source_expr.get_children()
+            if isinstance(element, ColumnClause)
+        )
+        return {element.name: element for element in elements}
 
     @_generative
     def update(self, values=None, where=None):
         # by default update all columns of target table that are
         # present in the source expression but exclude columns part
         # of the MERGE ON condition
         if values is None:
-            source_cols = self._get_source_cols()
             values = {}
-            for c in self._target_table.c:
-                if c not in self._on_columns and c.name in source_cols:
-                    values[c] = source_cols[c.name]
-        self._merge_update_values = ValuesBase(self._target_table, values, [])
+            src_columns = self._source_columns()
+            columns = (
+                column
+                for column in self.table.c
+                if column not in self._on_columns and column.name in src_columns
+            )
+            [values.update({column: src_columns[column.name]}) for column in columns]
```

Comment:
List comprehension should not be used , when the list is not assigned. In that case a normal for loop is usually recommended. An alternative in this case might be a dict or set merge in case values is a dict or set.